### PR TITLE
Add global config file, implement local and remote registry

### PR DESCRIPTION
**Pull request by minhqdao**
## tl;dr

- [x] Support global config file
- [x] Support local registry
- [x] Query and parse remote registry
- [x] Allow `namespace` and `v` (version) specifications for dependencies

## Global configuration file

This PR adds a global config file where you can define registry options:

```toml
[registry]
path = 'abc' # Location of the local registry (do not use with url or cache_path)
url = 'abc' # Can be used to download packages from a custom remote registry
cache_path = 'abc' # Cache location for downloaded remote packages
```

The default location of the global config file is `~/.local/share/fpm/config.toml` on Unix-based systems and `%APPDATA%\local\fpm\config.toml` on Windows.

Packages will be queried from the official registry if no config file exists or no `[registry]` subtable or `url` were specified.

The default `cache_path` will be set to `~/.local/share/fpm/dependencies` on Unix-based systems and `%APPDATA%\local\fpm\dependencies` on Windows if not specified.

Both absolute and relative paths can be resolved.

## Local Registry

Our goal is to have an official registry that acts as a reliable source for your dependencies. If you somehow prefer to store your dependencies locally, e.g. during development or to have your dependencies bundled with your code, you can define a directory that acts as your local registry in the global config file (https://github.com/fortran-lang/fpm/issues/321). `fpm` will then be able to search the local registry for dependencies listed in the manifest that are not `git` or direct `path` dependencies. It can search for the newest version as well as find a specific version.

The local registry needs to be structured in a way that `fpm` is able to find the correct package. A `namespace` must be provided and the package must be inside a folder that has the version number as its name. The manifest of an example package called "example" (`namespace`: "abc", `v`: "0.1.2") must be located at `<path>/abc/example/0.1.2/fpm.toml`.

## Remote Registry

If `url` isn't specified in the global config file, dependencies will the drawn from the official registry. In case you decide to use your own custom registry, follow [these docs](https://fpm-registry.onrender.com/apidocs/) to build your api endpoints. You can query the latest package (according to semantic versioning) or request a specific version. Packages, which are found, will be placed in the cache and are available across the system.

## Manifest Changes

`namespace` and `v` (version) are added as new keys to manifest dependencies. If you want to use the registry (either local and remote), a `namespace` must be provided for every package. It will allow using different packages that carry the same name. If you specify `v`, then `fpm` will look for that particular version. Without `v`, you will get the latest version.

Follow-ups:
- Support entering of a custom path to the global config file via the CLI.
- Support entering of a path or url to the registry via the CLI.
- Cache git dependencies as well.
- Handle constraints.
- Clear registry cache with `fpm clean --registry-cache`.
- Support packages in the cache that have the same version.
- Document.
- Implement `fpm publish`.

Relevant issues: https://github.com/fortran-lang/fpm/issues/551, https://github.com/fortran-lang/fpm/issues/814.

@awvwgk @everythingfunctional @gnikit @milancurcic @perazz @henilp105 @arteevraina

### CODE REVIEW

File: test/fpm_test/test_registry.f90 [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+contains
+
+   !> Collect unit tests.
+   subroutine collect_registry(tests)
+
+      !> Unit tests to collect.
+      type(unittest_t), allocatable, intent(out) :: tests(:)
+
+      tests = [ &
+      & new_unittest('no-tmp-folder', no_tmp_folder), &
```
Comment (everythingfunctional) [inline test/fpm_test/test_registry.f90:24]: It's somewhat better practice to have tests not depend on each other or depend on order of execution. I would recommend just handling this as part of the other tests. I.e. just put at the beginning of each test

```fortran
if (is_dir(tmp_folder)) call os_delete_dir(os_is_unix(), tmp_folder)
```

File: src/fpm_filesystem.F90 [hunk 2]
```diff
@@ -963,4 +963,36 @@ subroutine os_delete_dir(unix, dir, echo)
 
 end subroutine os_delete_dir
 
+!> Set path prefix to the local folder. Used for installation, registry etc.
```
Comment (everythingfunctional) [inline src/fpm_filesystem.F90:966]:
```suggestion
!> Determine the path prefix to the local folder. Used for installation, registry etc.
```

File: src/fpm_registry.f90 [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+         default_path_to_config_file = join_path(get_local_prefix(), 'fpm', 'config.toml')
+         if (exists(default_path_to_config_file)) then
+            path_to_config_file = default_path_to_config_file
+         end if
+      end if
+
+      ! Obtain registry settings from config file if it was found.
+      if (allocated(path_to_config_file)) then
+         allocate (reg_settings)
+         reg_settings%working_dir = path_to_config_file
```
Comment (everythingfunctional) [inline src/fpm_registry.f90:42]: Do you want this to be path to the file or the path to the folder containing the file?

**Review by everythingfunctional — COMMENTED** [review]
I think this is probably headed in the right direction. Just had a few comments/suggestions.

File: src/fpm_registry.f90 (hunk 3 above)
Comment (minhqdao) [reply src/fpm_registry.f90:42]: Oh, good catch! It's supposed to be the folder excluding the file!

Comment (minhqdao) [reply src/fpm_registry.f90]: I was wondering whether we'd want to work with the path to the folder or to the file here. I chose the file for now. 🤔

**Review by everythingfunctional — APPROVED** [review]

**Comment by minhqdao** [comment]
Maybe we can get one or two more opinions on this to get it merged soon. I kind of need it to continue working on the local registry. Also this is just the first step of a bigger feature so things might (and most probably will) change on the way. @fortran-lang/fpm

**Comment by minhqdao** [comment]
Will put this on draft, change it up and keep working on it a bit. 🙃

File: test/fpm_test/test_filesystem.f90 [hunk 4]
```diff
@@ -173,7 +174,86 @@ subroutine check_rmdir(error, path)
...
+
+        if (.not. is_absolute_path('~/', is_unix=.true.)) then
+            call test_failed(error, "Path '~/' is absolute")
+            return
+        end if
+
+        if (.not. is_absolute_path('~/', is_unix=.true.)) then
+            call test_failed(error, "Path '~/' is absolute")
+            return
+        end if
```
Comment (awvwgk) [inline test/fpm_test/test_filesystem.f90:210]: Those two check the same condition. What about cases like ``~user/``? Should we really recognize this kind of shell expansion or leave it to the shell to expand those values (passing ``~/`` via the command-line will be expanded by the shell to ``$HOME/`` and we never see the tilde).

File: src/fpm_command_line.f90 [hunk 5]
```diff
@@ -52,10 +53,28 @@ module fpm_command_line
...
+    !> Path to the global config file excluding the file name.
+    character(len=:), allocatable :: path_to_folder
+    !> Name of the global config file. The default is `config.toml`.
+    character(len=:), allocatable :: file_name
+    type(fpm_registry_settings), allocatable :: registry_settings
+contains
+    procedure :: full_path    
+end type
+
+type, extends(fpm_cmd_settings)   :: fpm_registry_settings
```
Comment (awvwgk) [inline src/fpm_command_line.f90:71]: I'm a bit confused here, the registry setting is not really a command line setting. It is also handled completely different from any other setting defined here.

I wonder whether it makes sense to actually have the settings here or move them into a new module to keep them logically separated.

File: src/fpm_os.F90 [hunk 6]
```diff
@@ -102,4 +102,21 @@ subroutine c_f_character(rhs, lhs)
...
+        character(len=:), allocatable :: start_dir
+  
+        call get_current_directory(start_dir, error)
+        if (allocated(error)) return
+        call change_directory(rel_path, error)
+        if (allocated(error)) return
+        call get_current_directory(abs_path, error)
+        if (allocated(error)) return
+        call change_directory(start_dir, error)
+        if (allocated(error)) return
```
Comment (awvwgk) [inline src/fpm_os.F90:119]: That's an interesting approach. Do we have a function from the C stdlib we could call instead to avoid changing the directory. Resolving a relative path to an absolute one is definitely not easy, especially when symbolic links are involved.

File: test/fpm_test/test_filesystem.f90 (hunk 4 above)
Comment (minhqdao) [reply test/fpm_test/test_filesystem.f90:210]: Yeah, maybe we should not treat `~/` as absolute. 🤔

Comment (minhqdao) [reply test/fpm_test/test_filesystem.f90:210]: But it's definitely easier to treat it as absolute for now. 😅

Comment (awvwgk) [reply test/fpm_test/test_filesystem.f90:210]: How would you classify `~\` on Windows? In Powershell this refers to your home, in CMD it does not.

Comment (minhqdao) [reply test/fpm_test/test_filesystem.f90:210]: We could automatically expand it to the %homepath% if we find it 🤔

File: src/fpm_os.F90 (hunk 6 above)
Comment (minhqdao) [reply src/fpm_os.F90:119]: Yeah, `~` really doesn't work this way. (`exists('~')` is `.false.` and `change_directory` to it also doesn't work). Gotta try `getcwd`, `realpath` or `GetFullPathName`.

File: src/fpm_command_line.f90 (hunk 5 above)
Comment (minhqdao) [reply src/fpm_command_line.f90:71]: We could at some point support entering the registry path via the CLI. Then it would behave similarly to the others. Is that wanted? If no, then yes, this should go somewhere else.

Comment (minhqdao) [reply src/fpm_command_line.f90:71]: No, I think I understand now. How about `fpm_global_settings`, should it also be moved to somewhere else?

File: src/fpm_os.F90 (hunk 6 above)
Comment (perazz) [reply src/fpm_os.F90:119]: In [this thread](https://fortran-lang.discourse.group/t/getting-a-full-path-name/4137/10) there are hints and some bits of code. It seems like some OS-specific operations are needed. i.e., use `GetFullPathName` on MS Windows and `getcwd` otherwise

**Comment by minhqdao** [comment]
`realpath` seems like the way to for Unix paths, and on Windows one could try [_fullpath](https://learn.microsoft.com/en-us/cpp/c-runtime-library/reference/fullpath-wfullpath?view=msvc-170) or [GetFullPathA](https://learn.microsoft.com/en-us/windows/win32/api/fileapi/nf-fileapi-getfullpathnamea). But I'll revert the changes now because I currently can't test it on Windows to get it running, and also it isn't necessarily part of this PR. I'm also getting a strange `malloc` error (only during tests) with `test_settings` enabled.

**Comment by perazz** [comment]
@minhqdao do you think this ready for review? I think it's a good idea that we define a finish line and limit the scope of this PR to the creation/handling of a local package cache only. In other words, I suggest you open a new PR and move there the last 3 tasks: 
- Download packages from a remote registry into the cache using either curl or wget.
- Test handling of downloaded packages from a remote registry.
- Handle constraints.

**Comment by minhqdao** [comment]
> @minhqdao do you think this ready for review? I think it's a good idea that we define a finish line and limit the scope of this PR to the creation/handling of a local package cache only. In other words, I suggest you open a new PR and move there the last 3 tasks:
> 
> * Download packages from a remote registry into the cache using either curl or wget.
> * Test handling of downloaded packages from a remote registry.
> * Handle constraints.

Like I said to you on the last meeting, `Download packages from a remote registry into the cache using either curl or wget.` has also been finished here. I obviously need to resolve conflicts and need to add some tests. You will be informed when the PR is ready.

**Comment by perazz** [comment]
Sure. Ping me when I can start reviewing. I also think it would be a good idea to have another one from the team on board, @henilp105 @arteevraina whose of you would be available to also review this PR?

**Comment by arteevraina** [comment]
> Sure. Ping me when I can start reviewing. This PR has been a lot of work and we want it merged soon and right. I also think it would be a good idea to have another one from the team on board, @henilp105 @arteevraina whose of you would be available to also review this PR?

Yeah, I can take some time to review this.

**Comment by minhqdao** [comment]
Ready for review 😅

**Review by perazz — APPROVED** [review]
Mostly stylistic comments but LGTM, thank you @minhqdao!

File: src/fpm/downloader.f90 [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+    ptr => cast_to_object(j_value)
+    if (.not. associated(ptr)) then
+      call fatal_error(error, "Error parsing JSON from '"//url//"'."); return
+    end if
+
+    json = ptr
+  end
+
+  subroutine get_file(url, tmp_file, error)
```
[the remark this answers is not in the document]
Comment (minhqdao) [reply src/fpm/downloader.f90:54]: If you only look at the procedure name, then yes, you are right. But this is a member procedure, meaning that it will, for example, be used as `downloader%get_file()`. Therefore I think it will always be clear that this procedure downloads a file from the web.

File: src/fpm.f90
[the remark this answers is not in the document]
Comment (minhqdao) [reply src/fpm.f90:1]: I can't see what you are referring to.

File: fpm.toml [hunk 8]
```diff
@@ -8,12 +8,16 @@ copyright = "2020 fpm contributors"
...
 [dependencies.toml-f]
 git = "https://github.com/toml-f/toml-f"
-rev = "aee54c5a480d623af99828c76df0447a15ce90dc"
+rev = "963082b9a1ef1833c2cd2d29dc037dbdb42d4b84"
 
 [dependencies.M_CLI2]
 git = "https://github.com/urbanjost/M_CLI2.git"
 rev = "7264878cdb1baff7323cc48596d829ccfe7751b8"
 
+[dependencies.jonquil]
```
[the remark this answers is not in the document]
Comment (minhqdao) [reply fpm.toml]: I first tried json-fortran but it's not compatible with fpm because of the many preprocessor commands. It doesn't work with `FPM_BOOTSTRAP`.

File: src/fpm/dependency.f90 [hunk 9]
```diff
@@ -452,25 +449,27 @@ subroutine resolve_dependency(self, dependency, root, error)
...
+      proj_dir = join_path(root, dependency%path)
+    else if (allocated(dependency%git)) then
+      proj_dir = join_path(self%dep_dir, dependency%name)
+      fetch = .not. exists(proj_dir)
+      if (fetch) then
+        call dependency%git%checkout(proj_dir, error)
+        if (allocated(error)) return
       end if
+    else
+      call get_global_settings(global_settings, error)
```
[the remark this answers is not in the document]
Comment (minhqdao) [reply src/fpm/dependency.f90:568]: You are right and this should be changed.

File: src/fpm/downloader.f90 (hunk 7 above)
[the remark this answers is not in the document]
Comment (minhqdao) [reply src/fpm/downloader.f90]: It's been changed here: https://github.com/fortran-lang/fpm/pull/817/commits/0c590794729fc8a859c9dfca2a9f5a886897ec75

File: src/fpm/dependency.f90 (hunk 9 above)
[the remark this answers is not in the document]
Comment (minhqdao) [reply src/fpm/dependency.f90]: It's been changed here: https://github.com/fortran-lang/fpm/pull/817/commits/0c590794729fc8a859c9dfca2a9f5a886897ec75

File: src/fpm_settings.f90 [hunk 10]
```diff
@@ -0,0 +1,231 @@
...
+  use fpm_environment, only: os_is_unix
+  use fpm_error, only: error_t, fatal_error
+  use fpm_toml, only: toml_table, toml_error, toml_stat, get_value, toml_load, check_keys
+  use fpm_os, only: get_current_directory, change_directory, get_absolute_path, &
+                    convert_to_absolute_path
+  implicit none
+  private
+  public :: fpm_global_settings, get_global_settings, get_registry_settings, official_registry_base_url
+
+  character(*), parameter :: official_registry_base_url = 'https://fpm-registry.onrender.com'
```
Comment (arteevraina) [inline src/fpm_settings.f90:13]: Can we put this base url out side of the source code in some environment variable and read if from there ?

**Review by arteevraina — APPROVED** [review]
Changes look fine to me overall. Except the one in which we hardcode the base url to the registry which is subject to change for sure.

File: src/fpm_settings.f90 (hunk 10 above)
Comment (minhqdao) [reply src/fpm_settings.f90:13]: You can specify a custom registry in the global config file if that is what you mean

Comment (minhqdao) [reply src/fpm_settings.f90:13]: Yes, the base url might not be final here, but it is also nothing that should be set by the user. I'm also a bit puzzled whether we should leave it like that. @awvwgk

**Comment by minhqdao** [comment]
Thanks everyone, I'll merge this. Next PRs will be shorter :)
